### DISCUSSION TITLE
问一下 2.x 使用 cq 码的问题

### DISCUSSION THREAD

**Discussion by Arondight** [takayama-lily/oicq#275] [Q&A] [answered]
因为没啥示例，所以来问问 @takayama-lily 

---

1.x 我是直接字符串硬编码的 cq 码，例如直接发 `"[CQ:image,file=base64://xxx]"` 过去， 2.x 不知道该怎么做，是发送 `segment.fromCqcode("[CQ:image,file=base64://xxx]")` 过去吗？

另外问一下 cq 码和文本混合的消息也能这么发码？

---

1.X 我去掉了 raw_message 中的 cq:at ，并且只使用了 raw_message ，2.x 应该怎么做，是需要 `msg.raw_message = msg.toCqcode()` 然后再做处理吗？

> 插件的参数 `msg` 的 [oicq](https://github.com/takayama-lily/oicq.git) 的数据结构改变。原本的 [CommonMessageEventData](https://github.com/takayama-lily/oicq/blob/39704008d8f00056d435ab8fa96c50f0b88c0568/index.d.ts#L453-L462) 中的 `message` 字段是一个可能包含[这些类型](https://github.com/takayama-lily/oicq/blob/39704008d8f00056d435ab8fa96c50f0b88c0568/index.d.ts#L178-L375)的数组，但是为了统一 `message` 和 `raw_message` 字段，在 `../src/utils/load.js` 中剔除了 [TextElem](https://github.com/takayama-lily/oicq/blob/39704008d8f00056d435ab8fa96c50f0b88c0568/index.d.ts#L188-L194) 之外的所有类型并只保留一个 `TextElem` 。除此之外，将 `raw_message` 和 [GroupMessageEventData](https://github.com/takayama-lily/oicq/blob/39704008d8f00056d435ab8fa96c50f0b88c0568/index.d.ts#L469-L479) 中仅存的 `TextElem` 进行了统一。最后 `GroupMessageEventData` 中新增了字段 `atMe: boolean` 来表示这条消息是否 `@` 了机器人。
>
> 除此之外， `raw_message` 和 `message[0].data.text` 依次去除了 `@` 机器人的 CQ 码、命令前缀 `config.prefixes` 和行首空格。

---

1. 我看当前源码中的 `toCqcode` 和 `fromCqcode` 都已经标记为 deprecated ，这一块未来还会有什么更改吗？
2. 另外我个人不太明白禁止直接使用 cq 码有啥好处，感觉怪麻烦的。

**Comment by takayama-lily** [comment, marked as answer]
你列出的CQ码转换的处理都没有问题。

1. 标记为`@deprecated `是不推荐使用的意思，今后应该会保留且不会再更改。
2. 废弃CQ码是因为原生对象表现力更强，使用方便，更易读（不用套一大堆转义字符）。你觉得麻烦可能是因为从之前过来习惯了。CQ码最初是为了简单表现一个元素（酷Q时代只有一个属性），之后没有进行良好设计一直往里加东西，变得越来越臃肿，可读性也急剧下降，已经与设计初衷相违背，性能下降倒是其次的。所以希望能更多使用原生对象来表现消息元素，与CQ码互转仅作为一种兼容手段。

**Reply by Arondight** [reply]
1. 不推荐的话，如果我不使用两个 api ，可以直接按照以前的文本处理直接处理、发送带 cq 的文本吗
2. 如果不行的话，这些转换支持混杂了 cq 码和普通文本的字符串码
3. 对于你废弃 cq 的表述我明白是要有一个可用的对象直接构造 cq，不需要硬编码 cq 了，可是我不明白为啥聊天文本需要 fromcq/tocq 的转换呢

**Reply by Arondight** [reply]
还有就是我比较好奇的事情，既然官方推荐进行 cq 转换，但是又标记了唯一能做这件事的 api 为 deprecated，这是后续准备有什么动作吗

**Reply by takayama-lily** [reply]
不能直接发送**带CQ码的文本**。
CQ码本来就是混在普通文本中的。
不推荐使用转换，推荐做法是完全抛弃**带CQ码文本**。
转换函数的存在只是为了过渡方便。不会有后续动作。今后也不会删除，CQ码可以一直作为一种纯字符串表示消息的格式。

**Reply by Arondight** [reply]
1. 那就是说其实 1.x 和 2.x 中 raw_message 没有任何变化，我该怎么用还可以怎么用，根本不需要 `msg.raw_message = msg.toCqcode()` 对吧？看你的描述似乎我就直接当做普通文本处理 `raw_message` 就可以了？
2. 抛弃带 cq 码的文本，本来我发送时传参的 `"[CQ:at,xxx]一段文字"` 变成 `segment.at("xxx") + "一段文字"` 就可以了吗？

**Reply by Arondight** [reply]
我看了下 oicq 当前的代码，之前我在 1.x 可以直接发一个 `@某人 一段文本 一个图片 一段文本 一个图片` 这种消息出去，不知道 2.x 要怎么做，似乎 segment.xx 是用来构造 msg.elem 的

**Reply by takayama-lily** [reply]
* 有变化，`raw_message`不是CQ码文本了，具体是什么建议打印一下
* 拼成数组 (详细参数类型以d.ts头文件为准)

**Reply by Arondight** [reply]
感谢回复，希望能够永久保留转换函数，这样看来最佳的使用方式就是 `msg.raw_message = msg.toCqcode()`，然后将 `msg.raw_message` 当做纯文本处理，最后 `client.sendGroupMsg(id, segment.fromCqcode(parsed_text))` 将消息发送出去。另外问一下 `sendGroupMsg` 和 `sendPrivateMsg` 会做永久保留吗？

**Comment by Arondight** [comment]
@takayama-lily 实际上这两个转换函数用起来是有问题的，没有完全支持做到 1.x 和 2.x 的消息转换，我稍微改了一下扔在自己项目里了 https://github.com/Arondight/Adachi-BOT/blob/master/src/utils/oicq.js

修改了 tocqcode ，text 不再转换为 cq 码，和 1.x 一样变成 cq 码和文本混杂的形式。上游可能是想尽可能做的规范一些，不过如果完全搞得这么规范，结果就是我一升级 oicq，我的 fork 炸了。1.x 中提供了这种混杂的串，自然也会有对应处理这种混杂串的代码，2.x 中的兼容函数却无法提供这种串了，有点不合逻辑，建议最好兼容一下。

**Reply by takayama-lily** [reply]
确实，text应该是文本，不应该在CQ里面
